### DISCUSSION TITLE
TLS zenoh-pico

### DISCUSSION THREAD

**Discussion by wimdewitte** [eclipse-zenoh/roadmap#73] [Ideas and new features]
It's great that zenoh version 0.7.0 now supports mTLS.
Do you consider to support this as well for zenoh-pico?
Especially for the upcoming European RED Delegated Act, it is a must have.
I understand this is a lot of work because of the supported systems; but I guess the most favorable platform is esp32.

**Comment by kydos** [comment]
We should probably look into adding support for WolfSSL, that is supported on a few embedded targets.

**Reply by cguimaraes** [reply]
We would need to understand if the WolfSSL license is compatible, since I have some doubts.
In any case, the platforms currently supported by Zenoh-Pico have support for TLS. It seems a better option for the long-term to go with something that is natively supported by each platform.

**Reply by josuah** [reply]
Tomorrow, some of WolfSSL will be at FOSDEM2023 in Brussels, Belgium.
https://www.wolfssl.com/join-us-fosdem-2023/

In case anyone from Zenoh is there, that would be some occasion to chat.

**Reply by cguimaraes** [reply]
Unfortunately, none of our team will be there.
In any case, if you meet them you can always point them to this thread.

**Comment by cguimaraes** [comment]
The support for TLS in Zenoh-Pico would be done just as any other new transport.
It does not seem an issue with the current list of supported platforms.

**Comment by andreaz98** [comment]
As of today, on ESP32 it is possible to use MbedTLS which supports SSL and TLS as well. If anyone want to create a support for TLS on zenoh-pico for ESP32 as an activity project for the university, where would be the best place to start things up?

**Reply by cguimaraes** [reply]
Currently it is quite simple to include new transports in Zenoh-Pico for any platform as the code is implementing several abstractions for it.
   1. As TLS would be a completely new transport, the first thing is to define it link type and along with its properties in it in the core code. Here, the best thing is to do a quick search for `Z_LINK_TCP` and just replicate whatever you find for your `Z_LINK_TLS` transport.
   2. Implement the platform specific functions. This is a PR for Serial support in ESP-IDF ( https://github.com/eclipse-zenoh/zenoh-pico/pull/143 ), but it represents what would need to be done for this point.
   3. Extend the integration tests with TLS.

If you follow this 3-step approach and you submit a PR, we would follow-up with a review and eventually merged it upstream.

**Reply by wimdewitte** [reply]
Creating new transports will create lots of duplicate code. Even if TLS is used, it remains TCP. My idea was to switch between secure vs non-secure on the lowest layer (network), thus when creating the socket. I've worked out a POC (not tested yet) for esp-idf. Please take a look at [https://github.com/wimdewitte/zenoh-pico/commit/d77b13035f1eb8d127fba878961fd5c33a50b2ac](url)

**Reply by cguimaraes** [reply]
Thanks for sharing your PoC. Also, if you decide to go beyond a PoC and actually implement the full TLS support feel free to do a pull request.

You can still implement TLS as a new transport and avoid some code duplication by partially reusing any function defined for TCP (if there is an overlapping). But from a very quick look on your PoC it does not to be the case. 
In any case, you have the option to enable / disable transports at compilation time.

Finally, TLS and TCP locators are not defined with the same scheme (`tls/10.0.0.1:7447` vs `tcp/10.0.0.1:7447`) and they might define a different set of properties and metadata.

**Reply by andreaz98** [reply]
I outlined a roadmap I'll keep up to date as I progress in the implementation. The purpose of this roadmap is to keep up to date those interested in the feature or willing to collaborate. The feature will cover the board `ESP32` and the framework `esp-idf`.

[edit] The roadmap has been moved here https://github.com/eclipse-zenoh/zenoh-pico/pull/153

**Reply by cguimaraes** [reply]
Can you share where are you implementing this feature?
In case we want to have a look and provide early feedback.

**Reply by andreaz98** [reply]
Sure I can!
With an early view I wrote the header at `zenoh-pico/link/config/tls.h` and the implementation is at `zenoh-pico/src/link/unicast/tls.c` where I reuse the functions:
* `__z_parse_port_segment_tls`
* `__z_parse_address_segment_tls`
* `_z_endpoint_tls_valid`

and I have to implement the functions (in the format `_z_f_link_*_tls`):
* open
* listen
* close
* free
* write
* write_all
* read
* read_exact
* get_link_mtu
* new_link_tls

There might be the need of a definition of new functions (hence an addition to the API) also for certificate handling.

**Reply by gabrik** [reply]
@andreaz98 I think @cguimaraes was thinking about where to find your code (i.e., the link to your fork).
It would also be good if you can do a draft PR already in Zenoh-Pico so that everyone can keep track/review the code.

**Reply by andreaz98** [reply]
Ok, sorry, I misunderstood the question. You can follow and review the code here https://github.com/andreaz98/zenoh-pico/tree/tls-implementation-esp32-esp-idf

This is the draft PR: https://github.com/eclipse-zenoh/zenoh-pico/pull/153
